### PR TITLE
[DLLM]Fast-dLLM-v2 support with HierarchyBlock algorithm for parallel decoding

### PR DESCRIPTION
**Pull request by SensenGao**
## Motivation

Huge thanks to @Hills-code, the author of Fast-dLLM v2, for the support and guidance throughout this integration.

Support [NVIDIA Fast-dLLM-v2](https://huggingface.co/Efficient-Large-Model/Fast_dLLM_v2_7B) with HierarchyBlock algorithm.

This PR continues the original work by @baonudesifeizhai (AR init for Fast-dllm-V2 in https://github.com/sgl-project/sglang/pull/15013) and @Monstertail (parallel decoding / HierarchyBlock algorithm in https://github.com/sgl-project/sglang/pull/17577). All of their commits are preserved with original authorship. I took over, rebased onto latest `main`, fixed a scheduler return-contract bug that blocked end-to-end runs, switched the tests to use the tokenizer chat template so the model sees the same input a production client would send, and re-validated the results.

We would like to make the Fast-dLLM-v2 the standard for integrating new dLLMs and their corresponding algorithms, especially to show:
- How do we implement the algorithms without aggressively changing the systems?
- How do we test the models and decoding algorithms?

## Modifications

- Init NVIDIA Fast-dLLM-v2 with full logits.
- Init HierarchyBlock algorithm for parallel generation.
- Add a GSM8k integration test for Fast-dLLM-v2 that applies the tokenizer chat template to every prompt.
- [Todo] support batching when rebase with the newest version after this PR: https://github.com/sgl-project/sglang/pull/14883

Next steps:
- Support batching for the HierarchyBlock algorithm. Token shifts will be challenging in batching because we need to map token from the last block with the dLLM requests in a batch, without breaking the schedule batch abstraction. CC @ClawSeven

## Accuracy Tests

Ran few-shot GSM8K evaluation (5-shot, 200 questions) on `Fast_dLLM_v2_7B` via the HierarchyBlock algorithm with:

**Server args (shared across all TP sizes)**

* `--trust-remote-code`
* `--mem-fraction-static 0.7`
* `--max-running-requests 1`
* `--attention-backend flashinfer`
* `--dllm-algorithm HierarchyBlock`

**Eval args**

* `num_shots=5, num_questions=200, max_new_tokens=512, parallel=128, temperature=0`
* Every prompt (few-shot + question) is wrapped in the tokenizer chat template before being sent to the server.
* Hardware: A100 80GB.

**Results (TP1)**

* Accuracy: **0.805**
* Invalid: **0.000**
* Latency: **295.251 s** (200 questions)
* Output throughput: **169.903 token/s**

```
Accuracy: 0.805
Invalid: 0.000
Latency: 295.251 s
Output throughput: 169.903 token/s
[tp1] metrics={'tp_size': 1, 'accuracy': 0.805, 'invalid': 0.0, 'latency': 295.25141421100125, 'output_throughput': 169.90265782148066, 'total_new_tokens': 50164}
```

**Results (TP2)**

* Accuracy: **0.770**
* Invalid: **0.000**
* Latency: **204.996 s** (200 questions)
* Output throughput: **244.444 token/s**

```
Accuracy: 0.770
Invalid: 0.000
Latency: 204.996 s
Output throughput: 244.444 token/s
[tp2] metrics={'tp_size': 2, 'accuracy': 0.77, 'invalid': 0.0, 'latency': 204.99613714800216, 'output_throughput': 244.44363048568965, 'total_new_tokens': 50110}
```

**Results (TP4)**

* Accuracy: **0.780**
* Invalid: **0.000**
* Latency: **162.577 s** (200 questions)
* Output throughput: **314.043 token/s**

```
Accuracy: 0.780
Invalid: 0.000
Latency: 162.577 s
Output throughput: 314.043 token/s
[tp4] metrics={'tp_size': 4, 'accuracy': 0.78, 'invalid': 0.0, 'latency': 162.57653282396495, 'output_throughput': 314.04286407856017, 'total_new_tokens': 51056}
```

To run the test, check: `test/srt/dllm/test_fast_dllm_v2.py` (classes `TestFastDLLMv2`, `TestFastDLLMv2TP2`, `TestFastDLLMv2TP4`).

## Benchmarking and Profiling

Included a bs=1 speed sanity benchmark in unit tests (`test_bs_1_speed`) using the same chat-template-formatted prompt across all TP sizes with `max_new_tokens=2048`. Each run decodes a single request end-to-end and reports output-token throughput.

| TP | new_tokens | latency (s) | speed (token/s) |
|----|-----------:|------------:|----------------:|
| 1  | 690        | 7.34        | **94.06**       |
| 2  | 539        | 4.10        | **131.51**      |
| 4  | 773        | 4.48        | **172.63**      |

Additionally, the GSM8K run reports **output throughput = 314.043 token/s** under TP4 with 128-way client concurrency (see Accuracy Tests).

## Speedup vs Standard AR Baseline (Qwen2.5-7B-Instruct)

To quantify the speedup from the HierarchyBlock algorithm, we benchmarked the standard autoregressive model **Qwen2.5-7B-Instruct** under identical server settings (`--mem-fraction-static 0.7`, `--max-running-requests 1`, `--attention-backend flashinfer`, chat template applied, A100 80GB).

**Qwen2.5-7B-Instruct Baseline Results**

| TP | Accuracy | Invalid | Latency (s) | Output Throughput (token/s) |
|----|----------|---------|------------:|----------------------------:|
| 1  | 0.825    | 0.000   | 401.554     | 96.358                      |
| 2  | 0.850    | 0.000   | 247.619     | 156.858                     |
| 4  | 0.830    | 0.000   | 155.042     | 248.495                     |

**Throughput Comparison: Fast-dLLM v2 vs Qwen2.5-7B-Instruct (AR baseline)**

| TP | AR Baseline Acc | Fast-dLLM v2 Acc | AR Baseline (token/s) | Fast-dLLM v2 (token/s) | Speedup |
|----|----------------:|-----------------:|----------------------:|-----------------------:|--------:|
| 1  | 0.825           | 0.805            | 96.358                | 169.903                | **1.76x** |
| 2  | 0.850           | 0.770            | 156.858               | 244.444                | **1.56x** |
| 4  | 0.830           | 0.780            | 248.495               | 314.043                | **1.26x** |

Fast-dLLM v2 with HierarchyBlock achieves **1.26x–1.76x** output throughput compared to standard autoregressive decoding, while maintaining comparable accuracy. The speedup is most pronounced at TP1 where the HierarchyBlock parallel decoding compensates most effectively for the single-GPU compute bottleneck.

To run the baseline test, check: `test/srt/dllm/test_qwen2_5_7b_instruct.py`.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

---
### CI States

Latest PR Test (Base): :x: [Run #27938993077](https://github.com/sgl-project/sglang/actions/runs/27938993077)
Latest PR Test (Extra): :x: [Run #27938992433](https://github.com/sgl-project/sglang/actions/runs/27938992433)

### CODE REVIEW

**Comment by Monstertail** [comment]
@btw616 @ClawSeven Could you please review the code? 
BTW, I am not sure whether we should keep test_fast_dllm_v2.py. In my impression,  SDAR deleted the similary test before merging .

File: python/sglang/srt/dllm/algorithm/hierarchy_block.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        self.token_shift = config.algorithm_config.get("token_shift", 1)
+
+        self.last_inherited_token = None
+
+    def run(
+        self,
+        model_runner: ModelRunner,
+        forward_batch: ForwardBatch,
+    ) -> Tuple[
+        Union[LogitsProcessorOutput, torch.Tensor], Optional[torch.Tensor], bool
```
[the remark this answers is not in the document]
Comment (btw616) [reply python/sglang/srt/dllm/algorithm/hierarchy_block.py]: Please fix this.

File: test/srt/dllm/test_fast_dllm_v2.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
+"""
```
Comment (btw616) [inline test/srt/dllm/test_fast_dllm_v2.py]: Currently, dllm tests should be placed in `test/registered/dllm/`.

File: test/srt/dllm/test_fast_dllm_v2.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        print(f"[{tag}] {metrics=}")
+
+        if is_in_ci():
+            write_github_step_summary(
+                f"### test_gsm8k (fast-dllm-v2, chat template) with {tag}\n"
+                f"accuracy={accuracy:.3f}, "
+                f"throughput={output_throughput:.2f} token/s\n"
+            )
+
+        self.assertGreater(metrics["accuracy"], 0.00)
```
Comment (btw616) [inline test/srt/dllm/test_fast_dllm_v2.py]: Comparing with zero doesn't help much; it would be better to compare with a more meaningful value.

File: test/srt/dllm/test_fast_dllm_v2.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        )
+        metrics = {
+            "tp_size": self.tp_size,
+            "accuracy": accuracy,
+            "invalid": invalid_rate,
+            "latency": total_latency,
+            "output_throughput": output_throughput,
+            "total_new_tokens": total_new_tokens,
+        }
+        print(f"[{tag}] {metrics=}")
```
Comment (btw616) [inline test/srt/dllm/test_fast_dllm_v2.py]: We can leverage the existing `sglang.test.run_eval` to simplify the test. You can refer to:
https://github.com/sgl-project/sglang/blob/7f8e7a913004c31774de9379e9c65d9a05fe5d6e/test/registered/dllm/test_llada2_mini.py#L59-L76

Comment (SensenGao) [reply test/srt/dllm/test_fast_dllm_v2.py]: Thanks a lot for the review! I've fixed the issues you pointed out.

File: python/sglang/srt/dllm/algorithm/hierarchy_block.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    ]:
+        total_len = len(forward_batch.input_ids)
+        block_mask = forward_batch.input_ids == self.mask_id
+        num_masked = block_mask.sum().item()
+        num_sub_blocks = total_len // self.sub_block_size
+        block_start = total_len - num_masked
+
+        # Detect new request (positions start from 0) and clear inheritance.
+        is_new_request = False
+        if hasattr(forward_batch, "positions") and forward_batch.positions is not None:
```
Comment (btw616) [inline python/sglang/srt/dllm/algorithm/hierarchy_block.py]: This check isn't necessary and could be removed.

File: python/sglang/srt/dllm/algorithm/hierarchy_block.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            if forward_batch.positions[0] == 0:
+                is_new_request = True
+                self.last_inherited_token = None
```
Comment (btw616) [inline python/sglang/srt/dllm/algorithm/hierarchy_block.py]: This makes the algorithm stateful and assumes that bs equals 1. It would be better to add a sanity check for that assumption.

File: python/sglang/srt/models/fast_dllm_v2.py [hunk 7]
```diff
@@ -0,0 +1,610 @@
+﻿# Copyright 2023-2024 SGLang Team
```
[the remark this answers is not in the document]
Comment (btw616) [reply python/sglang/srt/models/fast_dllm_v2.py]: As gemini-code-assist suggested above, it would be better to remove the BOM before `#`.

File: python/sglang/srt/dllm/algorithm/hierarchy_block.py (hunk 6 above)
Comment (SensenGao) [reply python/sglang/srt/dllm/algorithm/hierarchy_block.py]: Thanks for the review! I've applied the suggested changes, but I'm not fully confident the fix matches what you had in mind — could you take another look when you get a chance? Happy to iterate further if anything still needs adjusting.

**Comment by btw616** [comment]
LGTM

**Review by ClawSeven — APPROVED** [review]
LGTM

**Comment by ClawSeven** [comment]
Support the batching capability for fast-dllm-v2 in another PR. @SensenGao

**Comment by SensenGao** [comment]
@btw616  Hi, just a quick update. The pre-commit check has now been fixed. I noticed there are still a few failing tasks (such as the AMD ones). Do these need to be fixed as well?
